specs/vep-2448-vdk-run-logs-simplified-and-readable/README.md

```diff
@@ -366,10 +366,290 @@ or slowed down due to progress bar updates. This is really relevant for
+        """
+
+    def create_sub_tracker(self, title: str, total_iterations: int) -> ProgressTracker:
+        """
+        Create a new child tracker. The child tracker will point to its parent.
+        So any iterations added to the child tracker are automatically tracked by the parent as well.
+        And will update the parent automatically with any progress.
+        Any tracked metric is propagated to the parent as well
+        """
+
+    def track_metric(self, key: str, value: float):
+        """
+        The tracked metric must be cumulative since it's aggregated in the parent.
+        Non-cumulative metrics (like rate, latency, mean, median) will not provide result.
+        Example of cumulative metrics are  records processed, errors, time spent, number of requests, successful requests, failed requests.
+        """
+        self._metrics[key] = self._metrics.get(key, 0) + value
+        if self._parent:
+            self._parent.track_metric(key, value)
+
+    def get_metrics(self) -> Dict[str, int]:
+        """
+        Return dictionary with currently collected metrics about what is being tracked.
+        """
+```
+
+###### ProgressStrategy
+An abstract class in vdk-core that specifies the contract for strategy implementations in different environments.
+The basic logging strategy will be implemented in vdk-core while other strategies like CLI and notebook will be implemented in separate plugins like vdk-tqdm and vdk-jupyter.
+
+###### vdk-tqdm plugin
+
+For implementing CLI-based progress tracking TQDM (via vdk-tqdm plugin)
+It can provide capabilities for both Notebook (through tqdm.notebook package) and termianl (tqdm.tqdm package)
+
+In tqdm, implementing nested progress bars involves using the position and leave parameters.
+The position parameter sets which row the progress bar will appear in the console,
+while leave controls whether the progress bar stays visible after completion.
+
+Since trackers are never removed we can simply count all currently created trackers to set a position.
+
+##### Pseudo code
+
+```python
+import time
+from typing import List, Dict
+import threading
+
+MIN_TIME_BETWEEN_PROGRESS_UPDATE_SECONDS=5
+
+class ProgressTracker:
+    """
+    A class for tracking the progress of a task and its sub-tasks.
+    """
+
+    def __init__(self, title: str, parent: 'ProgressTracker', strategy: 'ProgressStrategy'):
+        """
+        :param title: The title of the task being tracked.
+        :param parent: The parent task tracker. None if this is the root task.
+        :param strategy: The strategy to use for tracking progress.
+        """
+        self._title = title
+        self._parent = parent
+        self._children: List['ProgressTracker'] = []
+        self._metrics: Dict[str, float] = {}
+        self._progress = 0
+        self._last_update_time = 0
+        self._last_updated_progress = 0
+        self._total_iterations = 0
+        self._strategy = strategy
```

Comment:
It's a bit confusing calling this a strategy. It's an implementation for the progress bar the specific tool will use (e.g. cli will use one kind of progress bar and notebooks will use a different kind). It's the strategy pattern, but we should just call it a progress bar.

Reply:
`ProgressStrategy` not named after the strategy design pattern. But more like this is a strategy for updating or sending the progress to the user . Using a progress bar would be one way in a local run, but we also would use just logging (for cloud use cases) without a progress bar. 

But yes, clearly ProgressStrategy is confusing. @gageorgiev  also noted that.  

Maybe `ProgressReporter`? `ProgressPresenter`? `ProgressNotifier`? 

Ideas